### PR TITLE
Answer:5

### PR DESCRIPTION
**Pull request by HosseinSalmanian**
Crud app refactored for pulling local state out of component using component store. Some other features added such as a global loading indicator for fetching todos list, local loading indicator for each todo item and an error message for each crud action which is performed on a todo item. some test also are wrote for todo store service.

### CODE REVIEW

File: apps/crud/src/app/services/todo.service.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+  public viewModel$ = this.select(
+    this.$todos,
+    this.$isLoading,
+    (todos, isLoading) => {
+      return {
+        todos,
+        isLoading,
+      };
+    }
+  );
```
Comment (tomalaforge) [inline apps/crud/src/app/services/todo.service.ts:39]: there is a new API where you can directly write : 
```ts
  public viewModel$ = this.select({
    todos$: this.$todos,
    isLoading$: this.$isLoading,
    });
```
And I like to add the debounce property to true on the viewModel only to update it only once.

File: apps/crud/src/app/services/todo.service.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      }),
+      concatMap((todoItem) =>
+        this.api.deleteTodo(todoItem.id).pipe(
+          map(() => false),
+          catchError(() => {
+            this.setItemErrorMessage(todoItem.id);
+            return of(true);
+          }),
+          tap((hasError) => !hasError && this.deleteItem(todoItem.id)),
+          finalize(() => this.stopProcessingItem(todoItem.id))
```
Comment (tomalaforge) [inline apps/crud/src/app/services/todo.service.ts:59]: you have the `tapResponse` operator which combine all of that.

**Comment by tomalaforge** [comment]
Very nice work and very complete solution. great job 💪⭐️

**Comment by HosseinSalmanian** [comment]
@tomalaforge  thank you for reviewing my pr.
your comments was very helpful.

**Pull request closed without merging**